test/test_ruby_parser.rb

```diff
@@ -185,6 +185,13 @@ def test_begin_else_return_value
     assert_syntax_error rb, "else without rescue is useless"
   end
 
+  def test_begin_rescue_ensure_no_bodies
+    rb = "begin\nensure\nend"
+    pt = s(:ensure, s(:nil).line(2)).line(2)
+
+    assert_parse rb, pt
+  end
```

Comment:
and added a bunch more tests w/ more coverage across all these crazy edge cases.